### PR TITLE
p2p: tighten up Router and add tests

### PR DESCRIPTION
**Pull request by erikgrinaker**
This cleans up the `Router` code and adds a bunch of tests. These sorts of systems are a real pain to test, since they have a bunch of asynchronous goroutines living their own lives, so the test coverage is decent but not fantastic. Luckily we've been able to move all of the complex peer management and transport logic outside of the router, as synchronous components that are much easier to test, so the core router logic is fairly small and simple.

This also provides some initial test tooling in `p2p/p2ptest` that automatically sets up in-memory networks and channels for use in integration tests. It also includes channel-oriented test asserters in `p2p/p2ptest/require.go`, but these have primarily been written for router testing and should probably be adapted or extended for reactor testing.

### CODE REVIEW

**Comment by erikgrinaker** [comment]
The race detector is triggering on Testify mocks accessing `context.Context` internals, which I believe is this bug:

https://github.com/stretchr/testify/issues/625

Will look into workarounds tomorrow.

**Review by tessr — APPROVED** [review]
This (huge) PR was (still) very readable! Nice work.

File: p2p/router.go [hunk 1]
```diff
@@ -28,83 +126,62 @@ type RouterOptions struct {
 	HandshakeTimeout time.Duration
 }
 
-// Validate validates the options.
+// Validate validates router options.
 func (o *RouterOptions) Validate() error {
 	return nil
```
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply p2p/router.go:131]: No idea, but possible. Could just remove it until it's needed too.

**Comment by erikgrinaker** [comment]
> This (huge) PR was (still) very readable! Nice work.

Thanks! Should ideally have split out some of the changes, but there were a bunch of fairly minor changes and I was short on time. Thanks for reviewing!

**Pull request merged**